tree-building/tree_test.go

```diff
@@ -0,0 +1,313 @@
+				{ID: 1},
+				{ID: 2},
+			},
+		},
+	},
+	{
+		name: "three nodes in reverse order",
+		input: []Record{
+			{ID: 2, Parent: 0},
+			{ID: 1, Parent: 0},
+			{ID: 0},
+		},
+		expected: &Node{
+			ID: 0,
+			Children: []*Node{
+				{ID: 1},
+				{ID: 2},
+			},
+		},
+	},
+	{
+		name: "more than two children",
+		input: []Record{
+			{ID: 3, Parent: 0},
+			{ID: 2, Parent: 0},
+			{ID: 1, Parent: 0},
+			{ID: 0},
+		},
+		expected: &Node{
+			ID: 0,
+			Children: []*Node{
+				{ID: 1},
+				{ID: 2},
+				{ID: 3},
+			},
+		},
+	},
+	{
+		name: "binary tree",
+		input: []Record{
+			{ID: 5, Parent: 1},
+			{ID: 3, Parent: 2},
+			{ID: 2, Parent: 0},
+			{ID: 4, Parent: 1},
+			{ID: 1, Parent: 0},
+			{ID: 0},
+			{ID: 6, Parent: 2},
+		},
+		expected: &Node{
+			ID: 0,
+			Children: []*Node{
+				{
+					ID: 1,
+					Children: []*Node{
+						{ID: 4},
+						{ID: 5},
+					},
+				},
+				{
+					ID: 2,
+					Children: []*Node{
+						{ID: 3},
+						{ID: 6},
+					},
+				},
+			},
+		},
+	},
+	{
+		name: "unbalanced tree",
+		input: []Record{
+			{ID: 5, Parent: 2},
+			{ID: 3, Parent: 2},
+			{ID: 2, Parent: 0},
+			{ID: 4, Parent: 1},
+			{ID: 1, Parent: 0},
+			{ID: 0},
+			{ID: 6, Parent: 2},
+		},
+		expected: &Node{
+			ID: 0,
+			Children: []*Node{
+				{
+					ID: 1,
+					Children: []*Node{
+						{ID: 4},
+					},
+				},
+				{
+					ID: 2,
+					Children: []*Node{
+						{ID: 3},
+						{ID: 5},
+						{ID: 6},
+					},
+				},
+			},
+		},
+	},
+}
+
+var failureTestCases = []struct {
+	name  string
+	input []Record
+}{
+	{
+		name: "root node has parent",
+		input: []Record{
+			{ID: 0, Parent: 1},
+			{ID: 1, Parent: 0},
+		},
+	},
+	{
+		name: "no root node",
+		input: []Record{
+			{ID: 1, Parent: 0},
+		},
+	},
+	{
+		name: "duplicate node",
+		input: []Record{
+			{ID: 0, Parent: 0},
+			{ID: 1, Parent: 0},
+			{ID: 1, Parent: 0},
+		},
+	},
+	{
+		name: "duplicate root",
+		input: []Record{
+			{ID: 0, Parent: 0},
+			{ID: 0, Parent: 0},
+		},
+	},
+	{
+		name: "non-continuous",
+		input: []Record{
+			{ID: 2, Parent: 0},
+			{ID: 4, Parent: 2},
+			{ID: 1, Parent: 0},
+			{ID: 0},
+		},
+	},
+	{
+		name: "cycle directly",
+		input: []Record{
+			{ID: 5, Parent: 2},
+			{ID: 3, Parent: 2},
+			{ID: 2, Parent: 2},
+			{ID: 4, Parent: 1},
+			{ID: 1, Parent: 0},
+			{ID: 0},
+			{ID: 6, Parent: 3},
+		},
+	},
+	{
+		name: "cycle indirectly",
+		input: []Record{
+			{ID: 5, Parent: 2},
+			{ID: 3, Parent: 2},
+			{ID: 2, Parent: 6},
+			{ID: 4, Parent: 1},
+			{ID: 1, Parent: 0},
+			{ID: 0},
+			{ID: 6, Parent: 3},
+		},
+	},
+	{
+		name: "higher id parent of lower id",
+		input: []Record{
+			{ID: 0},
+			{ID: 2, Parent: 0},
+			{ID: 1, Parent: 2},
+		},
+	},
+}
+
+func (n Node) String() string {
+	return fmt.Sprintf("%d:%s", n.ID, n.Children)
+}
+
+func TestMakeTreeSuccess(t *testing.T) {
+	for _, tt := range successTestCases {
+		actual, err := Build(tt.input)
+		if err != nil {
+			var _ error = err
```

Comment:
Is this to assert that the error is actually of type `error`? If so, prob better to write:

```go
if _, ok := err.(error); !ok {
    t.Fatal("err was not type error")
}
```
Still, seems like a bit of a strange thing to test given we have static typing here :)